From 6c31daf7fac81c022a768a379e5a82507da22268 Mon Sep 17 00:00:00 2001
From: Alexander Graf <agraf@suse.de>
Date: Mon, 11 Feb 2013 02:49:15 +0100
Subject: [PATCH 049/169] use tb linkage for conditional branches

---
 target-arm/translate-a64.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

Index: qemu-1.6.0+dfsg/target-arm/translate-a64.c
===================================================================
--- qemu-1.6.0+dfsg.orig/target-arm/translate-a64.c	2013-11-05 22:23:26.223408070 +0000
+++ qemu-1.6.0+dfsg/target-arm/translate-a64.c	2013-11-05 22:23:26.219408081 +0000
@@ -246,17 +246,16 @@
     gen_helper_cond(tcg_condmatch, pstate, tcg_cond);
     tcg_gen_brcond_i32(TCG_COND_EQ, tcg_condmatch, tcg_zero, no_match);
 
-    gen_a64_set_pc_im(addr);
-    tcg_gen_exit_tb(0);
+    gen_goto_tb(s, 0, addr);
 
     gen_set_label(no_match);
-    gen_a64_set_pc_im(s->pc);
+    gen_goto_tb(s, 1, s->pc);
 
     tcg_temp_free_i32(tcg_zero);
     tcg_temp_free_i32(tcg_cond);
     tcg_temp_free_i32(tcg_condmatch);
 
-    s->is_jmp = DISAS_JUMP;
+    s->is_jmp = DISAS_TB_JUMP;
 }
 
 static void handle_cb(DisasContext *s, uint32_t insn)
