From 129503fad984220ee50ee6d45dcd2b2bac56ad7a Mon Sep 17 00:00:00 2001
From: Alexander Graf <agraf@suse.de>
Date: Mon, 11 Feb 2013 02:51:03 +0100
Subject: [PATCH 050/169] use tb linkage for even more conditional branches

---
 target-arm/translate-a64.c | 13 ++++++-------
 1 file changed, 6 insertions(+), 7 deletions(-)

Index: qemu-1.6.0+dfsg/target-arm/translate-a64.c
===================================================================
--- qemu-1.6.0+dfsg.orig/target-arm/translate-a64.c	2013-11-05 22:23:26.419407563 +0000
+++ qemu-1.6.0+dfsg/target-arm/translate-a64.c	2013-11-05 22:23:26.411407583 +0000
@@ -282,13 +282,12 @@
     } else {
         tcg_gen_brcond_i64(TCG_COND_EQ, tcg_cmp, tcg_zero, no_match);
     }
-    gen_a64_set_pc_im(addr);
-    tcg_gen_exit_tb(0);
+    gen_goto_tb(s, 0, addr);
 
     gen_set_label(no_match);
-    gen_a64_set_pc_im(s->pc);
+    gen_goto_tb(s, 1, s->pc);
 
-    s->is_jmp = DISAS_JUMP;
+    s->is_jmp = DISAS_TB_JUMP;
 
     tcg_temp_free_i64(tcg_cmp);
     tcg_temp_free_i64(tcg_zero);
@@ -314,16 +313,16 @@
     } else {
         tcg_gen_brcond_i64(TCG_COND_EQ, tcg_cmp, tcg_mask, no_match);
     }
-    gen_a64_set_pc_im(addr);
+    gen_goto_tb(s, 0, addr);
     tcg_gen_exit_tb(0);
 
     gen_set_label(no_match);
-    gen_a64_set_pc_im(s->pc);
+    gen_goto_tb(s, 1, s->pc);
 
     tcg_temp_free_i64(tcg_cmp);
     tcg_temp_free_i64(tcg_mask);
 
-    s->is_jmp = DISAS_JUMP;
+    s->is_jmp = DISAS_TB_JUMP;
 }
 
 static void handle_cinc(DisasContext *s, uint32_t insn)
