From ef5c29c3f3cae0b3b09858c6e07578b58a46c9ef Mon Sep 17 00:00:00 2001
From: Alexander Graf <agraf@suse.de>
Date: Wed, 20 Feb 2013 16:34:36 +0100
Subject: [PATCH 070/169] fix sub corner case

---
 target-arm/helper-a64.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

Index: qemu-1.6.0+dfsg/target-arm/helper-a64.c
===================================================================
--- qemu-1.6.0+dfsg.orig/target-arm/helper-a64.c	2013-11-05 22:23:30.643396628 +0000
+++ qemu-1.6.0+dfsg/target-arm/helper-a64.c	2013-11-05 22:23:30.639396638 +0000
@@ -77,7 +77,8 @@
         pstate |= PSTATE_C;
     }
 
-    if ((s1 > 0 && s2 < 0 && sr < 0) || (s1 < 0 && s2 > 0 && sr > 0)) {
+    /* XXX check if this is the only special case */
+    if ((!a1 && a2 == 0x8000000000000000ULL) || (s1 > 0 && s2 < 0 && sr < 0) || (s1 < 0 && s2 > 0 && sr > 0)) {
         pstate |= PSTATE_V;
     }
 
@@ -101,7 +102,7 @@
         pstate |= PSTATE_C;
     }
 
-    if ((s1 > 0 && s2 < 0 && sr < 0) || (s1 < 0 && s2 > 0 && sr > 0)) {
+    if ((!a1 && a2 == 0x80000000ULL) || (s1 > 0 && s2 < 0 && sr < 0) || (s1 < 0 && s2 > 0 && sr > 0)) {
         pstate |= PSTATE_V;
     }
 
