From 686f25a6833335e2a90271b932b9f2f776d6826e Mon Sep 17 00:00:00 2001
From: Alexander Graf <agraf@suse.de>
Date: Thu, 21 Feb 2013 01:41:49 +0100
Subject: [PATCH 075/169] cut off upper 32bit in orri

---
 target-arm/translate-a64.c | 4 ++++
 1 file changed, 4 insertions(+)

Index: qemu-1.6.0+dfsg/target-arm/translate-a64.c
===================================================================
--- qemu-1.6.0+dfsg.orig/target-arm/translate-a64.c	2013-11-05 22:23:31.607394132 +0000
+++ qemu-1.6.0+dfsg/target-arm/translate-a64.c	2013-11-05 22:23:31.603394142 +0000
@@ -559,6 +559,10 @@
         break;
     }
 
+    if (is_32bit) {
+        tcg_gen_ext32u_i64(tcg_dst, tcg_dst);
+    }
+
     if (setflags) {
         gen_helper_pstate_add(pstate, pstate, tcg_dst, cpu_reg(31), tcg_dst);
     }
