From 58d586e5331d6c297d0806508ff745b9278d5170 Mon Sep 17 00:00:00 2001
From: Alexander Graf <agraf@suse.de>
Date: Thu, 28 Feb 2013 02:48:03 +0100
Subject: [PATCH 088/169] fix ldp with vector

---
 target-arm/translate-a64.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: qemu-1.6.0+dfsg/target-arm/translate-a64.c
===================================================================
--- qemu-1.6.0+dfsg.orig/target-arm/translate-a64.c	2013-11-05 22:23:34.159387526 +0000
+++ qemu-1.6.0+dfsg/target-arm/translate-a64.c	2013-11-05 22:23:34.159387526 +0000
@@ -2394,7 +2394,8 @@
             handle_ldarx(s, insn);
         }
         break;
-    case 0x0D:
+    case 0x0c:
+    case 0x0d:
         reserved(s, insn, 29, 1, 1);
         handle_stp(s, insn);
         break;
