From 39a301fc79cabef8705cf290e50e1376dbe712a4 Mon Sep 17 00:00:00 2001
From: Alexander Graf <agraf@suse.de>
Date: Fri, 1 Mar 2013 17:57:08 +0100
Subject: [PATCH 093/169] add mrs handling for ctr_el0

---
 target-arm/translate-a64.c | 5 +++++
 1 file changed, 5 insertions(+)

Index: qemu-1.6.0+dfsg/target-arm/translate-a64.c
===================================================================
--- qemu-1.6.0+dfsg.orig/target-arm/translate-a64.c	2013-11-05 22:23:35.231384751 +0000
+++ qemu-1.6.0+dfsg/target-arm/translate-a64.c	2013-11-05 22:23:35.227384761 +0000
@@ -401,6 +401,11 @@
         /* XXX this is wrong! */
         tcg_gen_ld32u_i64(cpu_reg(dest), cpu_env,
             offsetof(CPUARMState, vfp.xregs[ARM_VFP_FPSCR]));
+    } else if (op0 == 3 && op1 == 3 && op2 == 1 && crm == 0 && crn == 0) {
+        /* ctr_el0 */
+        /* CTR_EL0 [3:0] contains log2 of icache line size in words.
+           CTR_EL0 [19:16] contains log2 of dcache line size in words.  */
+        tcg_gen_movi_i64(cpu_reg(dest), 0x30003);
     } else {
         fprintf(stderr, "MRS: %d %d %d %d %d\n", op0, op1, op2, crm, crn);
         unallocated_encoding(s);
